Updated sensor abstraction

psigen opened the pull request:
This is a PR created by @christomaszewski to address the need for abstract sensors to represent internal as well as external sensors.

Commit:
Started making changes to sensor and motor architecture. Removed power
stuff from motor abstraction and created a peripheral abstraction to
represent the peripherals powered from the eboard (i.e. wifi, pump,
etc). Decoupled sensor ports from sensor abstraction; not all sensors
need a sensor port any more. Added a battery sensor and shell for the
IMU sensor.

Commit:
Moving abstract base sensor classes to Platypus.h/cpp. Classes that
can/should be instantiated are all places in Components.h/cpp. Removed
support for legacy GY26Compass module that was never used.

Commit:
Added initial IMU implementation based on Adafruit Unified sensor lib.
Doesn't recognize sensor or read data from it correctly yet.

Commit:
IMU sensor updates that should resolve issues with the standard Adafruit
library. Code currently discards readings when calibration values drop
below a certain level. Should probably initially drop readings until
first calibration is acquired then always report readings but warn if
readings are of poor quality.

Commit:
Updated sensor base class structure to include an ExternalSensor Base
class for sensors with an associated sensor port. Changed naming
convention to eliminate "channel" for sensors and replace with "id" for
all sensor and an addtional "port" parameter for external sensors. If no
port is specified for external sensors, the id will be used for the
port. This is done to maintain compatibility with old code but all new
code should specify a port number for external sensors.

Commit:
Added EmptySensor for initializing external ports that are not
connected to any sensor.

Commit:
Small changes in imu behavior - IMU constructor waits for calibration
before allowing code to procede. Should eventually indicate this to the
user via leds

Commit:
Small changes in firmware.ino defaults. Nothing critical here.

psigen commented on firmware/Board.h:
```diff
@@ -6,12 +6,16 @@
+    int ENABLE;
+    int CURRENT;
+  };
+
```
Can you explain a bit about how `Peripheral` and `Motor` relate?  It looks like some of the `Motor` functionality is abstracted into here.

christomaszewski replied:
I removed all the power control from the motor objects that we used to use when we powered the motors through the eboard and moved that into the peripheral objects so enable turns on the power and current reports the current draw of the peripheral. if we ever add some kind of current sensing for the motors we can current sensing back into the motor abstraction.

psigen commented on firmware/Board.h:
```diff
@@ -38,22 +42,32 @@ namespace board
+    { 48, A9 }  // Peripheral 1
+  };
+
   const size_t NUM_MOTORS = 2;
```
Does this always have to be `NUM_MOTORS == NUM_PERIPHERALS`?

christomaszewski replied:
No, motors has nothing to do with peripherals. At this point motors really just represent the servo outputs whereas the peripherals represent the power outputs we use to power the pump and wifi currently.

psigen commented on firmware/Platypus.cpp:
```diff
@@ -347,48 +365,156 @@ Sensor::Sensor(int channel)
+void AnalogSensor::offset(float offset)
 {
-  // Default to doing nothing on serial events. 
+  offset = offset_;
```
I think this is a bug and actually supposed to be `_offset = offset;`

christomaszewski replied:
Yea you're right. I'll make the change. We need to do another style cleanup because we have a variable naming mess.

psigen commented on firmware/Components.h:
```diff
@@ -68,92 +71,94 @@ namespace platypus
   private:
-    float scale_;
-    float offset_;
+    const int measurementInterval;
```
Might just rename this to `interval`, because it's the only timing related thing in here, but that's just a preference.  This is a fine name too.

psigen commented on firmware/Components.h:
```diff
@@ -68,92 +71,94 @@ namespace platypus
+    void onSerial();
+
   private:
     const int measurementInterval;
```
With `measurementInterval` appearing in multiple places, we may want to consider a mixin or composition to centralize the functionality in the future.

psigen commented on firmware/Components.cpp:
```diff
@@ -543,8 +460,8 @@ void AtlasPH::onSerial(){
-AtlasDO::AtlasDO(int channel) 
-  : Sensor(channel), SerialSensor(channel, 9600), measurementInterval(3000)
+AtlasDO::AtlasDO(int id, int port) 
+  : ExternalSensor(id, port), SerialSensor(id, port, 9600), measurementInterval(3000)
```
It might make sense to have `measurementInterval` be a constructor argument with a default value instead of being a constant hidden inside the initializer.

christomaszewski replied:
Yea makes sense. I would do a centralized measurementInterval but each sensor should have different default values. Also a different range of acceptable values.

psigen commented on firmware/Platypus.cpp:
```diff
@@ -309,27 +299,55 @@ void Motor::onLoop_(void *data)
-: channel_(channel)
+Sensor::Sensor(int id) : id_(id)
 {  
+  // Abstract Class - should not be instantiated
```
Instead of this, can we add a pure virtual function to this class so it _has_ to be abstract?

christomaszewski replied:
I think name() is pure virtual for sensor. The comment is just temporary/to remind people

psigen commented on firmware/Platypus.h:
```diff
@@ -85,26 +104,90 @@ namespace platypus
+
   };
   
+  class AnalogSensor : public ExternalSensor 
```
Does this class work now?

christomaszewski replied:
Not sure. I haven't tested it, just ported old code over to use the new class structure.

psigen commented on firmware/firmware.ino:
```diff
@@ -191,45 +191,47 @@ void setup()
   
   // Print header indicating that board successfully initialized
-  /*Serial.println(F("------------------------------"));
+  Serial.println(F("------------------------------"));
```
Can we change this to output a JSON structure of these same contents?  Then our phone server can tell what version firmware it is connected to on startup.

christomaszewski replied:
Yea can do that. What's the format that you want for that? Also how can we change the values so we can start incrementing the firmware versions without causing issues with the android version thing?

psigen replied:
https://github.com/platypusllc/arduino/blob/master/src/platypus/impl/Communication.cpp#L10

psigen replied:
The above link makes a JSON header that looks link:
```json
{
  "adk": {
    "company_name": "Platypus",
    "url": "http://senseplatypus.com",
    "accessory_name": "Platypus Controller",
    "version_number": "4.1.0"
  }
}
```

You can imagine adding a second key called `board` or something alongside `adk` which has a separate board version (e.g. `v4.2.0`) and firmware hash (e.g. 088566a8ef3703d684bbccce07a537d3546323a0) in it.

christomaszewski replied:
Should we change any of the strings used to instantiate the adk stuff in the firmware? We'll have to make updates to the phone server to support the new boards though.

christomaszewski replied:
This is what we currently have in the firmware:
char applicationName[] = "Platypus Server"; // the app on Android
char accessoryName[] = "Platypus Control Board"; // your Arduino board
char companyName[] = "Platypus LLC";
char versionNumber[] = "3.0";
char serialNumber[] = "3";
char url[] = "http://senseplatypus.com";

I remember updating the version number broke the phone's ability to communicate with the eboard.

psigen replied:
We should not update these numbers.  These are the _ADK protocol version_ numbers which are matched by the phone server.  That's why I made an `adk` section for the JSON I reported before.  These values have to stay in sync with the phone server.

We should instead have separate fields like `boardVersion` and `firmwareVersion` that we are updating and reporting.

psigen replied:
The one situation where we would change this number is if we were changing the device protocol in a non-backwards-compatible way.  At some point we should consider doing that so we can use minor revisions to indicate changes in protocol like requiring sensor initialization.

christomaszewski replied:
Should we tie the firmware version to the board version to start?

psigen replied:
No, let's just define it to be `4.0.0` to start.

psigen reviewed (commented):
Nice changes @christomaszewski.  Had a few questions.

Commit:
Updated variable naming to include underscore for member variables.
Changed measurement interval to constructor parameter with default
value. Added #define default measurement interval values.

christomaszewski committed:
Changes to support running on old Eboards. #define LEGACY in
platypus.cpp used to choose type of board. LED flashing may need fix on
new neopixel LEDS in the future. Reordered the sensor ports to match
pins in Board.h and convention where sensor_port 0 is not connected to a
hardware serial port.

christomaszewski committed:
Fixed sensor port issues. Added support for Adafruit GPS.

christomaszewski commented:
Merging this down to 4.2.0 because we have been using this firmware on the new eboards and keeping it separate from 4.2.0 is annoying for the different features we are implementing now.
